src/bridge.c

```diff
@@ -10,104 +22,48 @@ void ts_document_log_to_stderr(TSDocument *document) {
-	assert(node->data != NULL);
-	assert(outNode != NULL);
-	TSNode temp = ts_node_child(*node, index);
-	if (temp.data == NULL) {
-		printf("got broken child for index %ld\n", index);
-	}
-	assert(temp.data != NULL);
-	*outNode = temp;
-}
-
-void ts_node_p_named_child(const TSNode *node, size_t index, TSNode *outNode) {
-	assert(node != NULL);
-	assert(node->data != NULL);
-	assert(outNode != NULL);
-	TSNode temp = ts_node_named_child(*node, index);
-	if (temp.data == NULL) {
-		printf("got broken child for index %ld\n", index);
-	}
-	assert(temp.data != NULL);
-	*outNode = temp;
-}
-
-
-size_t ts_node_p_start_char(const TSNode *node) {
-	assert(node != NULL);
-	assert(node->data != NULL);
-	return ts_node_start_char(*node);
-}
-
-size_t ts_node_p_end_char(const TSNode *node) {
-	assert(node != NULL);
-	assert(node->data != NULL);
-	return ts_node_end_char(*node);
-}
-
-
-size_t ts_node_p_start_byte(const TSNode *node) {
-	assert(node != NULL);
-	assert(node->data != NULL);
-	return ts_node_start_byte(*node);
-}
-
-size_t ts_node_p_end_byte(const TSNode *node) {
-	assert(node != NULL);
-	assert(node->data != NULL);
-	return ts_node_end_byte(*node);
-}
-
-
-size_t ts_node_p_start_point_row(const TSNode *node) {
-	assert(node != NULL);
-	assert(node->data != NULL);
-	return ts_node_start_point(*node).row;
-}
-
-size_t ts_node_p_start_point_column(const TSNode *node) {
-	assert(node != NULL);
-	assert(node->data != NULL);
-	return ts_node_start_point(*node).column;
-}
-
-size_t ts_node_p_end_point_row(const TSNode *node) {
-	assert(node != NULL);
-	assert(node->data != NULL);
-	return ts_node_end_point(*node).row;
-}
-
-size_t ts_node_p_end_point_column(const TSNode *node) {
-	assert(node != NULL);
-	assert(node->data != NULL);
-	return ts_node_end_point(*node).column;
+static inline Node ts_node_elaborate(const TSDocument *document, TSNode node) {
+  return (Node){
+    .node = node,
+    .type = ts_node_type(node, document),
+    .startPoint = ts_node_start_point(node),
+    .endPoint = ts_node_end_point(node),
+    .startByte = ts_node_start_byte(node),
+    .endByte = ts_node_end_byte(node),
+    .namedChildCount = ts_node_named_child_count(node),
+    .childCount = ts_node_child_count(node)
+  };
+}
+
+void ts_document_root_node_p(TSDocument *document, Node *outNode) {
+  assert(document != NULL);
+  assert(outNode != NULL);
+  TSNode root = ts_document_root_node(document);
+  assert(root.data != NULL);
+  *outNode = ts_node_elaborate(document, root);
+}
+
+
+void ts_node_copy_child_nodes(const TSDocument *document, const TSNode *parentNode, Node *outChildNodes, size_t count) {
+  assert(document != NULL);
+  assert(parentNode != NULL);
+  assert(outChildNodes != NULL);
+  assert(count >= 0);
+  uint32_t maxCount = ts_node_child_count(*parentNode);
+  uint32_t max = maxCount <= count ? maxCount : count;
+  for (uint32_t i = 0; i < max; i++) {
+    outChildNodes[i] = ts_node_elaborate(document, ts_node_child(*parentNode, i));
```

Comment:
I’m a tiny bit bummed that we can’t `memcpy`, or even better, return a pointer to an already-populated buffer with all of the information in `Node` present. Even with all of the optimizations presently in place, simply copying node buffers using `ts_node_copy_named_child_nodes` (below) takes ~4.6% of our runtime; by contrast, `peek` on `Node` takes ~0.8%.

I’d like to consider alternatives for a future PR, however, as this is already a substantial improvement over our current behaviour.